### PR TITLE
[train] update Ray Train Data Loading and Preprocessing Guide

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

This PR contains an initial iteration of improving the Ray Train documentation for improving the `Data Loading and Preprocessing` page.

The main changes are:

1. Update the quickstart to be more focused on the data flow
    - Loading → Preprocessing → Splitting → Consuming
2. Add more details to the comparison between PyTorch data utilities.
3. Slightly reorganize/shuffle the content for readability purposes.

In a follow-up PR, we plan to update the examples and include more code snippets of the varying functionality.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 1]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
+Inputting and splitting data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Your preprocessed datasets can be passed into a Ray Train Trainer (e.g. :class:`~ray.train.torch.TorchTrainer`) through the ``datasets`` argument.
+
+The datasets passed into the Trainer's ``datasets`` can be accessed inside of the ``train_loop_per_worker`` run on each distributed training worker by calling :meth:`ray.train.get_dataset_shard`. 
+
+The default splitting behavior is as follows:
+
+- The ``"train"`` dataset is split (i.e. sharded) across the training workers. ``get_dataset_shard`` will return ``1/n`` of the dataset, where ``n`` is the number of training workers.
```
Comment (amogkam) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:218]: use :meth:`ray.train.session.get_dataset_shard` when applicable

Comment (matthewdeng) [reply doc/source/train/user-guides/data-loading-preprocessing.rst]: Oops good catch

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 2]
```diff
@@ -44,18 +49,21 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
-            # for creating the model and the training loop to train the model.
-            # See the Ray Train user guides for information such as how to report
-            # metrics or periodically save model checkpoints.
             def train_func(config):
                 model = nn.Sequential(nn.Linear(1, 1), nn.Sigmoid())
                 loss_fn = torch.nn.BCELoss()
                 optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
 
-                # Step 3: Access the dataset shard for the training worker via
+                # Step 4: Access the dataset shard for the training worker via
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:66]: I am thinking if we can simplify this `train_func`, it's kind of hard to follow now. Since we are discussing data loading here, we probably don't need to show how to train a model and report a checkpoint.

```python
def train_func()
    # Step x: ...
    train_ds = train.get_dataset_shard("train") # Sharded
    val_ds = train.get_dataset_shard("val") # Unsharded

    # Step y: ...
    for batch in ds.iter_torch_batches(batch_size=16):
        assert type(batch["x"]) == torch.Tensor
        assert type(batch["y"]) == torch.Tensor
        assert batch["x"].shape == (16,)
        assert batch["y"].shape == (16,)
```

Having some code like above should be enough?

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 3]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
+- For optimizing expensive preprocessing operations, see :ref:`Caching the preprocessed dataset <dataset_cache_performance>`.
+
+.. _train-datasets-input:
+
+Inputting and splitting data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Your preprocessed datasets can be passed into a Ray Train Trainer (e.g. :class:`~ray.train.torch.TorchTrainer`) through the ``datasets`` argument.
+
+The datasets passed into the Trainer's ``datasets`` can be accessed inside of the ``train_loop_per_worker`` run on each distributed training worker by calling :meth:`ray.train.get_dataset_shard`.
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:214]: Screenshot 2023-08-17 at 10 34 09 AM

Nit: seems to have too many hyperlinks here, in a single paragraph, shall we only link the api for the first time, and use ```` in the following occurrences?

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 4]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
 
-PyTorch Datasets are replaced by the :class:`Dataset <ray.data.Dataset>` abtraction, and the PyTorch DataLoader is replaced by :meth:`Dataset.iter_torch_batches() <ray.data.Dataset.iter_torch_batches>`.
+- To create a generic Iterable of batches, you can call :meth:`~ray.data.DataIterator.iter_batches`.
+- To create a replacement for a PyTorch DataLoader, you can call :meth:`~ray.data.DataIterator.iter_torch_batches`.
+
+For more details on how to iterate over your data, see :ref:`Iterating over data <iterating-over-data>`.
+
+.. _train-datasets-pytorch:
+
+Starting with PyTorch data
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:239]: Should we move this section under `Inputting and splitting data` section?

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 5]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
-Customizing how to split datasets
----------------------------------
+        **Option 1:** Convert your Hugging Face Dataset to a Ray Dataset and pass it into the Trainer via the ``datasets`` argument.
+        Inside your ``train_loop_per_worker``, you can access the dataset via :meth:`ray.train.get_dataset_shard`. 
+
+        For instructions, see :ref:`Ray Data for Hugging Face <loading_datasets_from_ml_libraries>`.
+
+        **Option 2:** Instantiate the Hugging Face Dataset directly in the ``train_loop_per_worker``.
+
+    .. tab:: LightningDataModule
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:291]: Can we move lightning to the second tab, so that the users know the "same logic" here is in the first tab (PyTorch Dataset and DataLoader)?

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 6]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
+     - :meth:`ray.data.Dataset.iter_torch_batches`
+
+
+For more details, see the following sections for each framework.
+
+.. tabs::
+
+    .. tab:: PyTorch Dataset and DataLoader
+
+        **Option 1:** Convert your PyTorch Dataset to a Ray Dataset and pass it into the Trainer via  ``datasets`` argument.
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:273]:
```suggestion
        **Option 1 (Ray Data):** Convert your PyTorch Dataset to a Ray Dataset and pass it into the Trainer via  ``datasets`` argument.
```

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 7]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
+        Inside your ``train_loop_per_worker``, you can access the dataset via :meth:`ray.train.get_dataset_shard`. 
+        You can convert this to replace the PyTorch DataLoader via :meth:`ray.data.DataIterator.iter_torch_batches`.
+         
+        For more details, see the :ref:`Migrating from PyTorch Datasets and DataLoaders <migrate_pytorch>`.
 
-For more details, see the :ref:`Ray Data PyTorch guide <migrate_pytorch>` and :ref:`Ray Data for HuggingFace and TensorFlow <loading_datasets_from_ml_libraries>`.
+        **Option 2:** Instantiate the Torch Dataset and DataLoader directly in the ``train_loop_per_worker``.
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:279]:
```suggestion
        **Option 2 (without Ray Data):** Instantiate the Torch Dataset and DataLoader directly in the ``train_loop_per_worker``.
```

Comment (woshiyyya) [reply doc/source/train/user-guides/data-loading-preprocessing.rst:279]: Is this clearer? otherwise the users might try to build ray data inside the `train_loop_per_worker`.

**Review by woshiyyya — CHANGES_REQUESTED** [review]

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 8]
```diff
@@ -177,25 +185,123 @@ Using Ray Data and Ray Train for distributed training on large datasets involves
...
+   * - `torch.utils.data.DataLoader <https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader>`_
+     - n/a
+     - :meth:`ray.data.Dataset.iter_torch_batches`
+
+
+For more details, see the following sections for each framework.
+
+.. tabs::
+
+    .. tab:: PyTorch Dataset and DataLoader
```
Comment (woshiyyya) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:271]: Should we also mention `ray.data.from_torch` here? If they want to quickly convert their small torch datasets.

File: doc/source/train/user-guides/data-loading-preprocessing.rst (hunk 3 above)
Comment (matthewdeng) [reply doc/source/train/user-guides/data-loading-preprocessing.rst:214]: @amogkam lmk your opinion here, another option would be to shorten to `get_dataset_shard()` and keep the link.

File: doc/source/train/user-guides/data-loading-preprocessing.rst (hunk 8 above)
Comment (matthewdeng) [reply doc/source/train/user-guides/data-loading-preprocessing.rst:271]: I didn't put it there because it's described in the `Migrating from PyTorch Datasets and DataLoaders` page that's linked

Comment (woshiyyya) [reply doc/source/train/user-guides/data-loading-preprocessing.rst:271]: Got it. Make sense:)

**Review by woshiyyya — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]
